HW0/P5/P5.txt

```diff
@@ -1 +1,15 @@
 # Your answers here.
+1. Given that the total number of pennies in each bag is already known, there are a total of 256 numbers to add, so 255 total addition operations. Since each operation takes 1 second, a total of 255 seconds would be necessary for a single cashier to add the totals of 256 bags.
+
+2. With 8 cashiers, each can take on 32 bags to count up, so each would require only 31 seconds to complete their counts. Once this is done, the 8 numbers need to be summed, which can also be slightly parallelized. If the cashiers paired up and one cashier from each pair added the sum of the pair's bags and this process continued until only one cashier remained with one final sum, it would take another 3 seconds to combine all the sums, so 34 seconds in total would be required.
+
+3. With an infinite number of employees but a finite number of bags (256), only a finite subset of employees (256 of them) would actually be useful. If each was assigned a bag and the same strategy as above was used to pair them off and compute the sums in parallel, only log_2(256) = 8 seconds would be required to count up all the bags.
+
+4. Based on the above result, log_2(N) seconds would be required to sum all the N bags by an infinite number of employees. Be sure to round the number up to account for the extra step required if the total number of bags is not an exact power of 2. So ceil(log_2(N)) seconds are required. This is reflected in the included graph.
+
+5. At every step of the way, before every computation is made, one instance of communication must take place. So every summation will now take 2 seconds (1 for communication and 1 for summation) such that the total time will now be 2*log_2(256) = 2*8 = 16 seconds to compute all bags.
```

Comment:
But if communication can happen only one on one, and each cashier needs two numbers to perform a sum. Shouldn't it take three seconds (two for sending the numbers and one for adding)?